gpflow/base.py

```diff
@@ -225,125 +190,11 @@ def assign(
         :param read_value: if True, will return something which evaluates to the new
             value of the variable; if False will return the assign op.
         """
-        unconstrained_value = self.validate_unconstrained_value(value, self.dtype)
-        return self._unconstrained.assign(
+        unconstrained_value = _validate_unconstrained_value(value, self.transform, self.dtype)
+        return self.unconstrained_variable.assign(
```

Comment:
The super class already has this implemented. Making use of it makes it more clear what is happening: 
1. validate that the unconstrained value exists.
2. assign it.

```suggestion
        return super().assign(value, use_locking, name, read_value)
```

Reply:
I think you're comment is right, the suggestion is wrong (I guess you know that though) - If you're going to do that, you can simply method the function, it'll inherit it from the parent.

I suppose what you're suggesting is
```
unconstrained_value = super().assign(value, use_locking, name, read_value)
return tf.debugging.assert_all_finite(unconstrained_value, message=message)
```
Which makes sense to me.

Reply:
In that case, the transformation will be performed 2 times. We redefine the method so that we could do extra steps before assigning value without sacrificing performance.

Reply:
Not sure I understand what you mean. The snippet @sam-willis posted seems correct to me?

Reply:
it will only be performed twice if you call `_validate_unconstrained_value`, which is why I suggested just using `tf.debugging.assert_all_finite`. You could make the message a constant, if you don't want code duplication.

Reply:
I think you can remove `_validate_unconstrained_value`, `_unconstrained_value` and `_constrained_value`, and just check that after either assignment or initialisation the unconstrained value is finite.

Reply:
I can't apply the Sam's suggestion because it changes the value first and then checks correctness, whereas it should be vice versa.

Reply:
ah I do see your point, maybe it's better to not change the value if it's invalid, although an exception is thrown, so I'm not sure how important it is. I guess if you catch the exception it makes recovering easier

Reply:
`_constrained_value` is unused, and technically you should always have the identity for the now prior, so you don't need the if in `_unconstrained_value`, but I think those are very minor.